src/Framework/Constraint/Count.php

```diff
@@ -65,7 +66,7 @@ protected function getCountOf($other)
                 return $this->getCountOfGenerator($iterator);
             }
 
-            $key   = $iterator->key();
+            $key   = ($iterator instanceof Iterator) ? $iterator->key() : null;
```

Comment:
> traversable = array or iterable follower, you need to ensure both are covered by tests

I don't think we're on the same page. The [iterable](https://wiki.php.net/rfc/iterable) type may be an array or Traversable. A [Traversable](http://php.net/manual/en/class.traversable.php) may be a userland class that implements [Iterator](http://php.net/manual/en/class.iterator.php) or [IteratorAggregate](http://php.net/manual/en/class.iteratoraggregate.php), but it may also be an internal class that implements Traversable directly and the [`get_iterator`](http://www.phpinternalsbook.com/classes_objects/iterators.html#the-get-iterator-handler) Zend object handler. This is touched upon in the following note from the Traversable documentation:

> Internal (built-in) classes that implement this interface can be used in a foreach construct and do not need to implement IteratorAggregate or Iterator.

Such a test case would require a Traversable instance that is neither an Iterator or an IteratorAggregate instance. Short of adding a dev dependency on a third-party PHP extension that provides an internal Traversable class, I am asking how you would propose testing this. Although the approach described in my last comment ("IteratorAggregate returns another IteratorAggregate") would technically exercise the changed LOC, it makes assumptions about the current implementation of `getCountOf()`, which I would argue is broken per my explanation below.

----

The `getCountOf()` function has an early case to handle Countable instances or array types, which is outside the scope of my change. I assume that is already tested.

This PR is modifying logic within the case for a Traversable instance. Within the Traversable case, the function checks for an IteratorAggregate and attempts to unwrap it *once* by calling [`IteratorAggregate::getIterator()`](http://php.net/manual/en/iteratoraggregate.getiterator.php). The function then checks for a Generator instance and handles that separately. After that point, the existing implementation makes two assumptions, which I believe are premature:

 1. It assumes that the object is an Iterator instance when it invokes the `key()` method. This fails to acknowledge that `IteratorAggregate::getIterator()` can return any Traversable instance.
 2. It assumes any Iterator can be rewound.

Ignoring the need for test coverage for a second, it was not my intention that this PR be merged as-is. The existing change is only a suggestion that demonstrates the first problem with the existing implementation. My goal was to have the questions in the OP answered before submitting any changeset for review (hence the "RFC" in the title).